### ISSUE TITLE
Load scripts on demand - already built in!?!

### ISSUE THREAD

**Issue by lodici** [magarena/magarena#118]
Please check that I am not hallucinating...

While giving some consideration to issue #117, I discovered if you comment out -
```
initializeEngine(reporter);
```
in ``MagicSystem.initialize(...)`` everything still works - this is the method that does all the heavy lifting at startup and causes the delay. Yet, I can run a test game, AI vs AI, DeckStrCal perfectly well without it. I can even run as normal and play a new or resume an existing duel and everything still works. It appears that somehow there is already a built-in mechanism for loading the card data on demand. How weird is that!

Only the Cards Explorer and Deck Editor are visibly affected but even they still work - as you load decks the card pool gradually increases. If we can fix these two screens (by loading all cards when these screens are accessed) then we can offer a massively improved startup time.

**Comment by firemind** [comment]
Yeah, I think I'm at least partially responsible for that one.

When I was working on the firemind duel runner I noticed that it was loading all the cards in existence for every duel. It had a major performance impact so like you I commented out the initialize call and I added a method to dynamically load cards by card name. This allows the firemind worker to only load the two deck lists. There was an issue though with cards that depended on other cards that were not part of the deck list (like double faced cards, tokens, etc.) so I also added lazy loading by name to the getCard method.

I think my initial code in magic/data/CardDefinitions got improved upon quite a bit but the basic functionality is still the same.

**Comment by lodici** [comment]
Well, muchos kudos to you, @firemind! :clap: 

I've had a bit more of a look and I reckon we can make this work without too much effort. We can just delay the calling of ``initializeEngine`` until Cards Explorer or Deck Editor is accessed. Will probably need to add some sort of "busy..please wait"  message because then you will be looking at the same kind of delay as now at startup.

As @huggybaby mentioned in the forum, if you are only interested in playing duels then this will have enormous benefit, not least of which is memory consumption...
__Before__
![screenshot00074](https://cloud.githubusercontent.com/assets/2443519/6793560/18929aa8-d1be-11e4-957f-444566f4ae7c.png)
__After__
![screenshot00072](https://cloud.githubusercontent.com/assets/2443519/6793568/2e4536a8-d1be-11e4-9084-bfb84cc8f02c.png)

**Comment by lodici** [comment]
@melvinzhang could you take a look at [this simple patch](https://gist.github.com/lodici/0233e42e702236dfe040#file-0001-only-load-full-set-of-card-scripts-if-cards-explorer-patch). I did not want to commit in case there is something I have missed. These changes load all remaining scripts when accessing the Cards Explorer or Deck Editor. Thanks.

**Comment by melvinzhang** [comment]
@lodici I see two potential issues:
1) MagicGameLog initialized when opening Deck Editor/Cards Explorer, should move it out of initializeEngine to initialize

2) CardDefinitions.addDefinition assumes that cards are never added twice, so that every call to allPlayableCardDefs.put increases the size of allPlayableCardDefs by one, and each card definition is assigned a unique index using allPlayableCardDefs.size(). This is no longer true, since when initializeEngine is called it will add cards that have already been added again. Need to prevent this to restore the assumption

**Comment by lodici** [comment]
ok, thanks - I will take another look tomorrow.

**Comment by melvinzhang** [comment]
Another consideration is random deck generation from deck profile. Without the cards loaded, there is not cards to choose from for random deck generation.

**Comment by lodici** [comment]
> 1) MagicGameLog initialized when opening Deck Editor/Cards Explorer, should move it out of initializeEngine to initialize

Never really understood why this is initialized on startup instead of on creation of ``MagicGame`` but either way you are right - it should not be part of opening Deck Editor/Cards Explorer.

> 2) CardDefinitions.addDefinition assumes that cards are never added twice, so that every call to allPlayableCardDefs.put increases the size of allPlayableCardDefs by one...

``put`` replaces the existing value, it does not create a new entry. Still, it might be more efficient to check first and skip the operation.

**Comment by lodici** [comment]
> put replaces the existing value, it does not create a new entry. Still, it might be more efficient to check first and skip the operation.

Actually, although the ``put`` is not a problem it's all the code that comes after it so an existence check is mandatory.

**Comment by lodici** [comment]
Any objections if I push the changes on a separate branch from which a pull request can be created instead of having to fork a brand new repository?

**Comment by lodici** [comment]
I have decided to just publish the update - it is a single small change commit so any problems it can always be rolled back easily.

**Comment by melvinzhang** [comment]
I think all the methods that return List<MagicCardDefinition> should load all cards instead of only getSpellCards. For example, download images doesn't work unless getAllPlayableCardDefs first load all the cards. Fixed in https://github.com/magarena/magarena/commit/fa859ee1d7762d71642ca90ea7bb946ca65e5467

Not sure about getMissingCards though, suspect it needs to load all cards too, since the way it knows a card is missing is if it is NOT in allPlayableCards.

**Comment by lodici** [comment]
You might as well rollback to a7bae9920a1a45d772ca4b0020cf523dcd1fae31. As far as I can see you have removed the card on-demand functionality and replaced with __all-cards__ on demand which fires at startup anyway except no progress is displayed in the splash dialog.

**Comment by melvinzhang** [comment]
My apologies, I wasn't aware that getDefaultPlayableCards and getAllPlayableCards are on the startup path. I was trying to fix the download images bug.

**Comment by melvinzhang** [comment]
There are two points these methods are called during startup
1. the tribal and ability mono deck generators (I think these we can convert to lazy load when they are actually used to generate decks).
2. MissingImagesAlertButton.getAlertCaption

**Comment by melvinzhang** [comment]
Doesn't seem to be an easy way around MissingImagesAlertButton, it is not compatible with lazy card loading. I'll work on the deck generators.

**Comment by melvinzhang** [comment]
Resolve deck generator causing all cards to load on startup in https://github.com/magarena/magarena/commit/dd4c720a772ef7f478519ae62e4020a2ffe20379

**Comment by melvinzhang** [comment]
Confirmed that if call to getAllPlayableCardDefs() is removed from CardDefinitions.isMissingImages, start game and play from prebuilt decks does not trigger a full card load.

**Comment by melvinzhang** [comment]
The number of cases where no full cards load takes place is rather limited (prebuilt decks, saved duel). The main mode of play is intended to be with random 40 card decks which requires full cards load. I'm thinking that we should always do a full load, but on a separate thread so it doesn't block the game UI from loading. Any calls to the cards would block while the loading thread is still running. This will also solve the missing images alert, it will simply block until the all the cards are loaded.

**Comment by lodici** [comment]
> Doesn't seem to be an easy way around MissingImagesAlertButton, it is not compatible with lazy card loading.

I don't think it is possible because card image files are given different names to the associated script file. If we actually saved the image file with the same name as the script then the download alerter could do a simple filename comparison without requiring to load the script in order to identify the image name. However ``MagicFileSystem.getImageFilename`` implies an image filename may have a numeric suffix so that might throw a fly in the ointment. Need to look into it further.

**Comment by lodici** [comment]
> I'm thinking that we should always do a full load, but on a separate thread so it doesn't block the game UI from loading.

I don't believe the main menu screen should be displayed until it is useable. Displaying it immediately then blocking as soon as an option is selected is pointless. It is also much more complicated than the current process with the splash. It is the norm in most games to expect a degree of waiting on startup so I am quite comfortable with the current splash screen/progress indicator and think we should stick with that for now if full load is required.

**Comment by melvinzhang** [comment]
Currently we are blocking when a full card list is required and then initiating the full load there and then. I'm proposing to initiate the load when the program start so that by the time it is needed for explorer or deck editor, most likely has already been fully loaded. 

The crucial difference is when the full load is initiated. The current method is to only initiate it when needed, but requiring a long wait. The pro is if the full list is never needed, less memory is used.

The proposed method is to always initiate a load of the full list in initalize, but not to wait for it, let it load in the background. When it is needed, there is less blocking to wait for it to all load. The cons is the full list is always loaded and taking up memory.

**Comment by melvinzhang** [comment]
I have an implementation of the background loading in my local repo. The code is simpler, the card loading is only triggered from one place (vs currently it can be triggered from several places). The other place that require full card data first call waitForAllCards and block in this method until loading is complete, if loadings is already done, waitForAllCards return immediately.

**Comment by melvinzhang** [comment]
I'm only proposing the background load for initializeEngine, the main menu screen is usable and there is no blocking if playing a duel with prebuilt decks or restoring a saved duel, same as what we have now.

The improvement is less waiting when starting the card explorer. Instead of triggering initializeEngine when user open's card explorer and then waiting for the load to complete. We start initializeEngine in the background as part of the initialize method.

**Comment by lodici** [comment]
Which is all well and good but currently it will always do a full load because of the image download check at startup. So this will need to be disabled first.

**Comment by melvinzhang** [comment]
If we go with the background card def loading, then the image download check is not an issue, since it will automatically block until the backround loading is completed. There is also the matter of image_updated property the card check should take note of which requires loading the contents of the card script.

**Comment by lodici** [comment]
> the main menu screen is usable and there is no blocking if playing a duel with prebuilt decks or restoring a saved duel, same as what we have now.

Maybe not common specs these days, but on my Acer Aspire 110 across the network it can sometimes take well over a minute to startup while it loads the cards. With the new code, it would more or less display the main menu straightaway. If I resume a game, will it block? If I start a new duel using random pre-built decks will it block? If not then it is totally worth it just for these use-cases. 

There is still a problem that the startup waiting time is now offloaded onto actions performed in the UI instead of at a single point (and reported via the splash) so from my UI POV it is more work since I will have to come up with a way of observing the background card loading and displaying a wait message. Which is fine, but I have other stuff outstanding for 1.60 as well.

**Comment by melvinzhang** [comment]
> There is still a problem that the startup waiting time is now offloaded onto actions performed in the UI instead of at a single point

Yes, I think this is a problem as currently since we could trigger a load from several different points in the UI. The background loading would solve this as the UI thread never does any loading, it is done in the background thread. The background loading implementation has the UI thread blocking when it needs the data (vs actually doing the loading itself), but we can also provide a method in MagicSystem for the UI to query if loading is completed without blocking if we want to display a waiting message.

**Comment by lodici** [comment]
> but we can also provide a non-blocking method in MagicSystem for the UI to query if loading is completed if we want to display a waiting message.

preferably with a progress indication.

**Comment by melvinzhang** [comment]
> from my UI POV it is more work since I will have to come up with a way of observing the background card loading and displaying a wait message. 

There will be some waiting on opening the card explorer whether we do background card def loading  as suggested or the current on the spot loading when creating card explorer. So I take it this is not an argument against background card loading. 

> preferably with a progress indication.
Easily done with background loading. As a start, it will be the last progress message from ProgressReported.

**Comment by lodici** [comment]
> There will be some waiting on opening the card explorer whether we do background card def loading as suggested or the current on the spot loading when creating card explorer. So I take it this is not an argument against background card loading.

Not at all but in my example above with the new background loading, the minute or so of waiting is now transferred to the action of opening the Card Explorer so I have to display something in place of the progress message currently displayed in the splash since that it is far too long being busy without notifying the user what is going on. Its a useability thing.

**Comment by melvinzhang** [comment]
background loading released in 1.60
